### ISSUE TITLE
Zeroing Memory

### ISSUE THREAD

**Issue by aparent** [open-quantum-safe/liboqs#48]
We should be zeroing memory that is used to store secret data in our free functions.  It turns out that it is pretty difficult to do this in a way that is portable and guaranteed not to be optimized away.  We could do something a bit ugly as show below (from libsodium): 

```
#ifdef HAVE_WEAK_SYMBOLS
__attribute__ ((weak)) void
_sodium_memzero_as_a_weak_symbol_to_prevent_lto(void * const pnt, const size_t len)
{
    unsigned char *pnt_ = (unsigned char *) pnt;;
    size_t         i = (size_t) 0U;

    while (i < len) {
        pnt_[i++] = 0U;
    }
}
#endif

void
sodium_memzero(void * const pnt, const size_t len)
{
#ifdef _WIN32
    SecureZeroMemory(pnt, len);
#elif defined(HAVE_MEMSET_S)
    if (len > 0U && memset_s(pnt, (rsize_t) len, 0, (rsize_t) len) != 0) {
        abort(); /* LCOV_EXCL_LINE */
    }
#elif defined(HAVE_EXPLICIT_BZERO)
    explicit_bzero(pnt, len);
#elif HAVE_WEAK_SYMBOLS
    _sodium_memzero_as_a_weak_symbol_to_prevent_lto(pnt, len);
#else
    volatile unsigned char *volatile pnt_ =
        (volatile unsigned char * volatile) pnt;
    size_t i = (size_t) 0U;

    while (i < len) {
        pnt_[i++] = 0U;
    }
#endif
}
```

**Comment by mofojed** [comment]
OPENSSL just makes the pointer to the memset function volatile to avoid compiler optimization:

```
/*
 * Pointer to memset is volatile so that compiler must de-reference
 * the pointer and can't assume that it points to any function in
 * particular (such as memset, which it then might further "optimize")
 */
typedef void *(*memset_t)(void *, int, size_t);

static volatile memset_t memset_func = memset;

void OPENSSL_cleanse(void *ptr, size_t len)
{
    memset_func(ptr, 0, len);
}
```
Reference: https://github.com/openssl/openssl/blob/master/crypto/mem_clr.c

Would that be sufficient?

OpenSSL then has further optimizations as well, they have platform specific assembler that overrides the OPENSSL_cleanse function for optimizing further, eg: https://github.com/openssl/openssl/blob/master/crypto/x86_64cpuid.pl#L196

**Comment by aparent** [comment]
That looks OK for now. After reading some discussions online it does not seem clear that the solution is future proof though.

Windows has a function specifically for this (`SecureZeroMemory`) so we might as well use that when it is available.  We could add something like the below:

```
#include<string.h>

#if defined(WINDOWS)

#include<windows.h>

void OQS_cleanse(void *ptr, size_t len)
{
    SecureZeroMemory(pnt, len);
}

#else

typedef void *(*memset_t)(void *, int, size_t);

static volatile memset_t memset_func = memset;

void OQS_cleanse(void *ptr, size_t len)
{
        memset_func(ptr, 0, len);
}

#endif
```

**Comment by mofojed** [comment]
I like that approach.

I wrote some benchmarking testing different zeroing methods: https://github.com/open-quantum-safe/liboqs/compare/master...mofojed:zero

I have not run on a windows machine yet, so haven't compared with the `SecureZeroMemory` method. The other methods were all pretty much the same, with the exception of the sodium fallback method, which is much, much slower. We should avoid that one if possible :) Benchmark results below.

I also wrote a test for correctness, to verify that memory is actually being zero'ed out. The tests pass (including verifying a case where memset gets optimized out and the secret is still there), though I'm not certain this kind of test would capture all possible optimizations.

```
================================================================================
Benchmark Tests: volatile
================================================================================
Started at 2016-11-08 23:32:32
Operation                           Iterations  Total time (s)  Time(us): mean      pop. stdev    Cycles: mean      pop. stdev
small block                            7231821           1.000           0.138           0.482             310             299
large block                               1565           1.000         639.157          96.632         1784851          269883
Ended at 2016-11-08 23:32:34

================================================================================
Benchmark Tests: volatile2
================================================================================
Started at 2016-11-08 23:32:34
Operation                           Iterations  Total time (s)  Time(us): mean      pop. stdev    Cycles: mean      pop. stdev
small block                            7094546           1.000           0.141           0.481             317             905
large block                               1597           1.000         626.314          70.410         1748996          196629
Ended at 2016-11-08 23:32:36

================================================================================
Benchmark Tests: volatile3
================================================================================
Started at 2016-11-08 23:32:36
Operation                           Iterations  Total time (s)  Time(us): mean      pop. stdev    Cycles: mean      pop. stdev
small block                            7233185           1.000           0.138           0.369             308             315
large block                               1583           1.000         631.888          85.838         1764565          239685
Ended at 2016-11-08 23:32:39

================================================================================
Benchmark Tests: sodium
================================================================================
Started at 2016-11-08 23:32:39
Operation                           Iterations  Total time (s)  Time(us): mean      pop. stdev    Cycles: mean      pop. stdev
small block                             761482           1.000           1.313           1.114            3596            2883
large block                                198           1.004        5070.879         539.690        14165098         1507452
Ended at 2016-11-08 23:32:41
```

**Comment by dstebila** [comment]
Thanks guys, great work on figuring out and testing a good solution.

I suggest that we pick one of volatile/2/3/sodium and incorporate it into liboqs, along with Windows version and memset_s.  I don't think we need test_mem in liboqs; once we are happy withthe code, hopefully it will remain untouched.  

I was thinking that maybe we should do an src/common directory with a corresponding common.h, and put the OQS_MEM_cleanse in that.  I have ideas on a few other things to put in there as well, eventually, such as the UNUSED macro for flagging unused attributes, and some basic debugging macros.

**Comment by aparent** [comment]
Would the following implementation work? If so I can create the common directory and add it.

```C
#if defined(WINDOWS)

#include<windows.h>
void OQS_MEM_cleanse(void *ptr, size_t len){
    SecureZeroMemory(pnt, len);
}

#elif defined(HAVE_MEMSET_S)

#include<string.h>
void OQS_MEM_cleanse(void *const pnt, const size_t len) {
	if (0U < len && memset_s(pnt, (rsize_t) len, 0, (rsize_t) len) != 0) {
		abort(); /* LCOV_EXCL_LINE */
	}
}

#else 

#include<string.h>
typedef void *(*memset_t)(void *, int, size_t);
static volatile memset_t memset_func = memset;
void OQS_MEM_cleanse(void *ptr, size_t len){
        memset_func(ptr, 0, len);
}

#endif
```

**Comment by dstebila** [comment]
Looks good to me.  Mike, any comments?

With the same pull request, could you go through the rest of the codebase and put in these cleanse calls everywhere memory that could contain a secret value is freed?

**Comment by dstebila** [comment]
Close issue: merged PR #54.
